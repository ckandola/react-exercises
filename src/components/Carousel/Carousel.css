:root {
    /* primary/main color */
    --clr-primary-5: hsl(205, 78%, 60%);
    --clr-grey-3: hsl(209, 34%, 30%);
    /* grey used for paragraphs */
    --clr-grey-5: hsl(210, 22%, 49%);
    --clr-grey-8: hsl(210, 31%, 80%);

    --clr-white: #fff;
    --transition: all 0.3s linear;
    --radius: 0.25rem;
    --dark-shadow: 0 5px 15px rgba(0, 0, 0, 0.2);
    --max-width: 1170px;
}

.carousel-title {
    text-align: center;
    margin-bottom: 2rem;
}

.carousel-title h2 {
    display: flex;
    align-items: center;
    justify-content: center;
    font-weight: 500;
}
.carousel-title span {
    font-size: 0.85em;
    color: var(--clr-primary-5);
    margin-right: 1rem;
    font-weight: 700;
}

.carousel-section {
    width: 90vw;
    margin: 5rem auto;
    max-width: var(--max-width);
}

.carousel-section-center {
    margin: 0 auto;
    margin-top: 4rem;
    width: 80vw;
    /* have to have a height */
    height: 450px;
    max-width: 800px;
    text-align: center;
    position: relative;
    display: flex;
    overflow: hidden;
}

.carousel-person-img {
    border-radius: 50%;
    margin-bottom: 1rem;
    width: 150px;
    height: 150px;
    object-fit: cover;
    border: 4px solid var(--clr-grey-8);
    box-shadow: var(--dark-shadow);
}

.carousel-article h4 {
    text-transform: uppercase;
    color: var(--clr-primary-5);
    margin-bottom: 0.25rem;
}

.carousel-title {
    text-transform: capitalize;
    margin-bottom: 0.75rem;
    color: var(--clr-grey-3);
}

.carousel-text {
    max-width: 35em;
    margin: 0 auto;
    margin-top: 2rem;
    line-height: 2;
    color: var(--clr-grey-5);
}

.carousel-icon {
    font-size: 3rem;
    margin-top: 1rem;
    color: var(--clr-primary-5);
}

.carousel-prev,
.carousel-next {
    position: absolute;
    top: 200px;
    transform: translateY(-50%);
    background: var(--clr-grey-5);
    color: var(--clr-white);
    width: 1.25rem;
    height: 1.25rem;
    display: grid;
    place-items: center;
    border-color: transparent;
    font-size: 1rem;
    border-radius: var(--radius);
    cursor: pointer;
    transition: var(--transition);
}

.carousel-prev:hover,
.carousel-next:hover {
    background: var(--clr-primary-5);
}
.carousel-prev {
    left: 0;
}
.carousel-next {
    right: 0;
}
@media (min-width: 800px) {
    .carousel-text {
        max-width: 45em;
    }
    .carousel-prev,
    .carousel-next {
        width: 2rem;
        height: 2rem;
        font-size: 1.5rem;
    }
}
.carousel-article {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    opacity: 0;
    transition: var(--transition);
}

.carousel-article-activeSlide {
    opacity: 1;
    transform: translateX(0);
}
.carousel-article-lastSlide {
    transform: translateX(-100%);
}
.carousel-article-nextSlide {
    transform: translateX(100%);
}
