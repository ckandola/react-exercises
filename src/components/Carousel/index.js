// from https://youtu.be/a_7Z7C_JCyo?t=8088

import React, {useState, useEffect} from 'react';
import {FiChevronRight, FiChevronLeft} from 'react-icons/fi';
import {FaQuoteRight} from 'react-icons/fa';
import './Carousel.css';
import people from './data';

const Carousel = () => {
    const [ix, setIx] = useState(0);

    useEffect(() => {
        let slider = setInterval(() => {
            setIx(ix === people.length - 1 ? 0 : ix + 1);
        }, 5000);
        return () => clearInterval(slider);
    }, [ix])

    return (
        <section className="carousel-section">
            <div className="carousel-title">
                <h2>
                    <span>/ </span>Reviews
                </h2>
            </div>
            <div className="carousel-section-center">
                {people.map((person, index) => {
                    const {id, image, name, title, quote} = person;

                    let position = index === ix ? 'activeSlide' : 'nextSlide';
                    if (index === ix - 1 || (ix === 0 && index === people.length - 1)) {
                        position = 'lastSlide';
                    }
                    return (
                        <article className={`.carousel-${position}`} key={id}>
                            <img src={image} alt={name} className="carousel-person-img"/>
                            <h4>{name}</h4>
                            <p className="carousel-title">{title}</p>
                            <p className="carousel-text">{quote}</p>
                            <FaQuoteRight className="carousel-icon" />
                        </article>
                    );
                })};
                <button 
                    className="carousel-prev" 
                    onClick={() => setIx(ix === 0 ? people.length - 1 : ix - 1)}>
                        <FiChevronLeft />
                </button>
                <button 
                    className="carousel-next" 
                    onClick={() => setIx(ix === people.length - 1 ? 0 : ix + 1)}>
                        <FiChevronRight />
                </button>
            </div>
        </section>
    );
};

export default Carousel;
