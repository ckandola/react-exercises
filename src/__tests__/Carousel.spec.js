import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Carousel from '../components/Carousel';

describe('Carousel tests', () => {
    it('matches the snapshot', () => {
        const wrapper = shallow(<Carousel />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('changes the carousel contents every 5 seconds', () => {
        jest.useFakeTimers();
        const wrapper = shallow(<Carousel />);
        const firstPerson = wrapper.find('.carousel-article-activeSlide').getElement().props.children[0].props.alt;
        jest.advanceTimersByTime(5000);
        const nextPerson = wrapper.find('.carousel-article-activeSlide').getElement().props.children[0].props.alt;
        expect(firstPerson).not.toBe(nextPerson);
        jest.useRealTimers();
    });

    it('advances to next review when right arrow is clicked', () => {
        const wrapper = shallow(<Carousel />);
        expect(wrapper.find('.carousel-article-activeSlide').getElement().props.children[0].props.alt).toBe("maria ferguson");
        wrapper.find('.carousel-next').simulate('click');   // 2
        expect(wrapper.find('.carousel-article-activeSlide').getElement().props.children[0].props.alt).toBe('john doe');
        wrapper.find('.carousel-next').simulate('click');   // 3
        wrapper.find('.carousel-next').simulate('click');   // 4
        wrapper.find('.carousel-next').simulate('click');   // 1
        expect(wrapper.find('.carousel-article-activeSlide').getElement().props.children[0].props.alt).toBe('maria ferguson');
    });

    it('returns to previous review when left arrow is clicked', () => {
        const wrapper = shallow(<Carousel />);
        expect(wrapper.find('.carousel-article-activeSlide').getElement().props.children[0].props.alt).toBe("maria ferguson");
        wrapper.find('.carousel-prev').simulate('click');
        expect(wrapper.find('.carousel-article-activeSlide').getElement().props.children[0].props.alt).toBe('susan andersen');
        wrapper.find('.carousel-prev').simulate('click');
        expect(wrapper.find('.carousel-article-activeSlide').getElement().props.children[0].props.alt).toBe('peter smith');
    });
});
