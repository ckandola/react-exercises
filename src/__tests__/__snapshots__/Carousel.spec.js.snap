// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel tests matches the snapshot 1`] = `
<section
  className="carousel-section"
>
  <div
    className="carousel-title"
  >
    <h2>
      <span>
        / 
      </span>
      Reviews
    </h2>
  </div>
  <div
    className="carousel-section-center"
  >
    <article
      className="carousel-article carousel-article-activeSlide"
      key="1"
    >
      <img
        alt="maria ferguson"
        className="carousel-person-img"
        src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1595959131/person-2_ipcjws.jpg"
      />
      <h4>
        maria ferguson
      </h4>
      <p
        className="carousel-title"
      >
        office manager
      </p>
      <p
        className="carousel-text"
      >
        Fingerstache umami squid, kinfolk subway tile selvage tumblr man braid viral kombucha gentrify fanny pack raclette pok pok mustache.
      </p>
      <FaQuoteRight
        className="carousel-icon"
      />
    </article>
    <article
      className="carousel-article carousel-article-nextSlide"
      key="2"
    >
      <img
        alt="john doe"
        className="carousel-person-img"
        src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1586883417/person-3_ipa0mj.jpg"
      />
      <h4>
        john doe
      </h4>
      <p
        className="carousel-title"
      >
        regular guy
      </p>
      <p
        className="carousel-text"
      >
        Gastropub sustainable tousled prism occupy. Viral XOXO roof party brunch actually, chambray listicle microdosing put a bird on it paleo subway tile squid umami.
      </p>
      <FaQuoteRight
        className="carousel-icon"
      />
    </article>
    <article
      className="carousel-article carousel-article-nextSlide"
      key="3"
    >
      <img
        alt="peter smith"
        className="carousel-person-img"
        src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1595959121/person-1_aufeoq.jpg"
      />
      <h4>
        peter smith
      </h4>
      <p
        className="carousel-title"
      >
        product designer
      </p>
      <p
        className="carousel-text"
      >
        Drinking vinegar polaroid street art echo park, actually semiotics next level butcher master cleanse hammock flexitarian ethical paleo.
      </p>
      <FaQuoteRight
        className="carousel-icon"
      />
    </article>
    <article
      className="carousel-article carousel-article-lastSlide"
      key="4"
    >
      <img
        alt="susan andersen"
        className="carousel-person-img"
        src="https://res.cloudinary.com/diqqf3eq2/image/upload/v1586883334/person-1_rfzshl.jpg"
      />
      <h4>
        susan andersen
      </h4>
      <p
        className="carousel-title"
      >
        the boss
      </p>
      <p
        className="carousel-text"
      >
        Marfa af yr 3 wolf moon kogi, readymade distillery asymmetrical seitan kale chips fingerstache cloud bread mustache twee messenger bag. 
      </p>
      <FaQuoteRight
        className="carousel-icon"
      />
    </article>
    <button
      className="carousel-prev"
      onClick={[Function]}
    >
      <FiChevronLeft />
    </button>
    <button
      className="carousel-next"
      onClick={[Function]}
    >
      <FiChevronRight />
    </button>
  </div>
</section>
`;
